packages/upload-api/src/blob/replicate.js

```diff
@@ -0,0 +1,188 @@
+      const authRes = await Validator.claim(Assert.location, [claim], {
+        authority: context.id,
+        ...invContext
+      })
+      if (authRes.error) {
+        return Server.error(/** @type {API.InvalidReplicationSite} */ ({
+          name: 'InvalidReplicationSite',
+          message: `location commitment validation error: ${authRes.error.message}`
+        }))
+      }
+
+      // validate location commitment is for the digest we want to replicate
+      const claimDigest = 'multihash' in claim.capabilities[0].nb.content
+        ? claim.capabilities[0].nb.content.multihash
+        : Digest.decode(claim.capabilities[0].nb.content.digest)
+      if (!equals(claimDigest.bytes, digest.bytes)) {
+        return Server.error(/** @type {API.InvalidReplicationSite} */ ({
+          name: 'InvalidReplicationSite',
+          message: `location commitment blob (${base58btc.encode(claimDigest.bytes)}) does not reference replication blob: ${base58btc.encode(digest.bytes)}`
+        }))
+      }
+
+      const selectRes = await router.selectReplicationProviders(
+        claim.issuer,
+        nb.replicas,
+        digest,
+        nb.blob.size,
+        { exclude: existingReplicas.map(r => DID.parse(r.provider)) }
+      )
+      if (selectRes.error) {
+        return selectRes
+      }
+
+      const allocRes = await Promise.all(selectRes.ok.map(async (candidate) => {
+        const candidateDID = candidate.did()
+        const confRes = await router.configureInvocation(candidate, {
+          can: BlobReplica.allocate.can,
+          with: candidateDID,
+          nb: {
+            blob: nb.blob,
+            space: DID.parse(space),
+            site: nb.site,
+            cause: invocation.cid
+          }
+        })
+        if (confRes.error) {
+          return confRes
+        }
+
+        const { connection, invocation: allocInv } = confRes.ok
+
+        const receipt = await allocInv.execute(connection)
+        const task = receipt.ran
+        if (!Delegation.isDelegation(task)) {
+          throw new Error('expected receipt ran to be a delegation')
+        }
+
+        // record the invocation and the receipt, so we can retrieve it later
+        // when we get a blob/replica/transfer receipt in ucan/conclude
+        const message = await Message.build({
+          invocations: [task],
+          receipts: [receipt],
+        })
+        const messageWriteRes = await agentStore.messages.write({
+          source: await Transport.outbound.encode(message),
+          data: message,
+          index: [...AgentMessage.index(message)],
+        })
+        if (messageWriteRes.error) {
+          return messageWriteRes
+        }
+
+        const addRes = await replicaStore.add({
+          space,
+          digest,
+          provider: candidateDID,
+          status: receipt.out.error ? 'failed' : 'allocated',
+          cause: task.cid
+        })
+        return addRes.error ? addRes : receipt
+      }))
```

Comment:
will setting the status here have the same effect as `replicaStore.setStatus`? Asking since that method exists on the `replicaStore` but isn't used yet. Will its usage land as part of the seperate PR to add `ucan/conclude<blob/replica/transfer>` implementation?

Reply:
Yes correct - when we receive the `blob/replica/transfer` receipt we will update the status to `transferred`.